File: frogment/src/main/java/ch/grze/frogment/frogment/AbstractStateAwareFrogment.java
package ch.grze.frogment.frogment;

import ch.grze.frogment.State;

//still in Java due to lack of interop: https://youtrack.jetbrains.com/issue/KT-19415
public abstract class AbstractStateAwareFrogment<S extends State> extends AbstractFrogment implements StateAwareFrogmentInterface<S> {
    private StateAwareFrogmentComponent stateAwareFrogmentComponent;

    @Override
    public StateAwareFrogmentComponent getStateAwareFrogmentComponent() {
        return stateAwareFrogmentComponent;
    }

    @Override
    public void setStateAwareFrogmentComponent(StateAwareFrogmentComponent component) {
        stateAwareFrogmentComponent = component;
    }
}

File: frogment/src/main/java/ch/grze/frogment/frogment/AbstractStateAwareFrogment.kt
package ch.grze.frogment.frogment

import ch.grze.frogment.State
import ch.grze.frogment.StateAware

abstract class AbstractStateAwareFrogment<S : State> private constructor(override val stateAwareFrogmentComponent: StateAwareFrogmentComponent<S>) :
        AbstractFrogment(),
        StateAwareFrogmentInterface<S>,
        StateAware<S> by stateAwareFrogmentComponent
{
    constructor() : this(StateAwareFrogmentComponent<S>())

    init {
        stateAwareFrogmentComponent.stateAwareFrogment = this
    }
}
